### ISSUE TITLE
route parameter does not escape  "/"

### ISSUE THREAD

**Issue by kongfujianlong** [Kong/unirest-java#141]
======== background =======
I have a RESTful API looks like "GET /product/productID" productID is a route parameter.

I build the GetRequest like below:
GetRequest request = Unirest.get("http://localhost/product/{id}").routeParam("id", product_id);

======== problem =======
Assume product_id = "id/id"

If I print request.getUrl(), I get "id%2Fid". However, when the request get to my http server (nginx), it becomes "product/id/id"

Basically I think routeParam itself does convert "/" to percentage encoding, but the underlying httpclient implementation may altered this behavior.

Could you suggest any fix?

**Comment by hutchic** [comment]
clarifying:
you want `http://localhost/product/id/id` but it's giving you `http://localhost/product/id%2Fid` ?

**Comment by kongfujianlong** [comment]
I want http://localhost/product/id%2Fid, but it's giving me http://localhost/product/id/id

request.getUrl() prints me the correct string (http://localhost/product/id%2Fid), but when the request is sent, http server receives http://localhost/product/id/id

**Comment by froginvasion** [comment]
Link to this issue: #137 

I investigated, and URL encoding of route parameters is generally not very good implemented. It uses the `URLParamEncoder` class instead of relying on existing encoding utilities. I also fail to see which encoding this class applies, let alone that is is just not working correctly.

Using query parameters however does use a different way of encoding, namely by calling ´URLEncoder.encode("value", "UTF-8")`. The fact that this encoding is not variable, and that its encoding differs totally from encoding route parameters, are two red flags for me.

@kongfujianlong I recommend you do not use `routeParam` anymore. If you want to use unirest for now, then encode the route parameter yourself instead of relying on Unirest to do it for you.

I hope this gets fixed soon, it is a huge shortcoming and potential problem I just uncovered in our rest client. The only quick fix that I can propose is by replacing the ´URLParamEncoder` call to a call to `URLEncoder.encode`, and then at least it will use the same encoding method as query parameters.

**Comment by kongfujianlong** [comment]
Thanks. The problem is, even if I encode the URL by myself, the unirest still somehow modifies my URL (decode some of the encoded characters before sending out the request).

**Comment by kongfujianlong** [comment]
Using URLEncoder for URL encoding is probably not a good idea. URLEncoder is supposed to be used for  HTML form encoding, which has slightly different rule than the URL encoding.

Please see here http://stackoverflow.com/questions/14321873/java-url-encoding-urlencoder-vs-uri

**Comment by froginvasion** [comment]
@kongfujianlong you're right, it isn't. But I haven't found a good way to do it in java either. :(

**Comment by ryber** [comment]
Hello, this should be resolved or addressed in Unirest 2. Please see the [upgrade guide](https://github.com/Kong/unirest-java/blob/master/UPGRADE_GUIDE.md) for important changes.
